elytron/WFLY-16793-oidc-identity-propagation.adoc

```diff
@@ -0,0 +1,186 @@
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+* A new `virtual-security-domain` resource will be introduced in the `elytron` subsystem.
+The new resource will make it possible to outflow security identities established by virtual
+security domains to other security domains. The new `virtual-security-domain` resource will have
+two attributes:
+
+** `name` - This is the runtime name of a deployment associated with a virtual security domain (e.g.,
+`DEPLOYMENT_NAME.ear`, a deployment that has a subdeployment that is secured using OIDC).
+
+** `outflow-security-domains` - This is the list of `security-domains` that security identities from
+the virtual security domain should be automatically outflowed to. This will be a required attribute
+and must be non-empty.
+
+** As an example, a `virtual-security-domain` could be added as follows:
+[source]
+----
+/subsystem=elytron/virtual-security-domain=DEPLOYMENT_NAME.ear:add(outflow-security-domains=...)
+----
+
+* We also need the ability to specify the virtual security domains that should be trusted by a `security-domain`.
+Currently, when a virtual security domain is created, the `SecurityDomain` service that is provided has a
+name like `DEPLOYMENT_NAME.ear.security-domain.virtual`. So although it provides a `SecurityDomain`, the
+service name is formatted differently from a normal `security-domain` that's been configured.
+We could update this so that the service name for the virtual case is of the usual form `org.wildfly.security.security-domain.DEPLOYMENT_NAME.ear`.
+That way, a separate `trusted-virtual-security-domains` attribute would not be needed. Instead, the
+existing `trusted-security-domains` attribute for a `security-domain` could be used since we'd be able to
+determine the service name even for a virtual security domain in the usual way.
+** e.g., `trusted-security-domains=DEPLOYMENT_NAME.ear` would indicate that this `security-domain`
+should trust the virtual security domain associated with the `DEPLOYMENT_NAME.ear` deployment.
+
+=== Nice-to-Have Requirements
+
+Since we are now introducing a new `virtual-security-domain` resource, we can easily add some `security-domain`
+attributes in addition to `outflow-security-domains` since the ability to add configuration for virtual
+security domains is something that we have seen users asking for both on the WildFly user forum and in
+JIRA (e.g., see https://issues.redhat.com/browse/WFLY-17312[WFLY-17312] and https://issues.redhat.com/browse/WFLY-17333[WFLY-17333]).
+
+As a starting point, we could add the following attributes:
+
+* outflow-anonymous
+* pre-realm-principal-transformer
+* post-realm-principal-transformer
+* principal-decoder
+* evidence-decoder
+* role-decoder
+* role-mapper
+* permission-mapper
+* security-event-listener
```

Comment:
Just wondering, are some "identity propagation" events emitted? If it's the case, on which side?

Reply:
For identity propagation, the `ServerAuthenticationContext#importIdentity` method gets called to import the identity into the target domain (see https://github.com/wildfly-security/wildfly-elytron/blob/1.x/auth/server/base/src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java#L1630). If the identity can be propagated successfully, the target domain will emit a `RealmIdentitySuccessfulAuthorizationEvent`. Otherwise, it will emit a `RealmIdentityFailedAuthorizationEvent`.

Reply:
Note that for virtual security domains, the usual security events will also get emitted upon successful / failed authentication.

Reply:
Thank you @fjuma.